core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -0,0 +1,199 @@
+    idleInboxes.put(endpoint, inbox)
+    afterUpdateInbox(inbox)
+    endpointRef
+  }
+
+  def getRpcEndpointRef(endpoint: RpcEndpoint): RpcEndpointRef = endpointToEndpointRef.get(endpoint)
+
+  def getRpcEndpointRef(name: String): RpcEndpointRef = nameToEndpoint.get(name).endpointRef
+
+  // Should be idempotent
+  def unregisterRpcEndpoint(name: String): Unit = {
+    val endpointPair = nameToEndpoint.remove(name)
+    if (endpointPair != null) {
+      val inbox = endpointToInbox.remove(endpointPair.endpoint)
+      if (inbox != null) {
+        inbox.stop()
+        afterUpdateInbox(inbox)
+      }
+      endpointToEndpointRef.remove(endpointPair.endpoint)
+    }
+  }
+
+  def stop(rpcEndpointRef: RpcEndpointRef): Unit = {
+    unregisterRpcEndpoint(rpcEndpointRef.name)
+  }
+
+  /**
+   * Send a message to all registered [[RpcEndpoint]]s.
+   * @param message
+   */
+  def broadcastMessage(message: BroadcastMessage): Unit = {
+    val iter = endpointToInbox.values().iterator()
+    while(iter.hasNext) {
+      val inbox = iter.next()
+      postMessageToInbox(inbox, message)
+    }
+  }
+
+  def postMessage(message: RequestMessage, callback: RpcResponseCallback): Unit = {
+    val receiver = nameToEndpoint.get(message.receiver.name)
+    if (receiver != null) {
+      val inbox = endpointToInbox.get(receiver.endpoint)
+      if (inbox != null) {
+        val rpcCallContext =
+          new RemoteNettyRpcCallContext(
+            nettyEnv, inbox.endpointRef, callback, message.senderAddress, message.needReply)
+        postMessageToInbox(inbox,
+          ContentMessage(message.senderAddress, message.content, message.needReply, rpcCallContext))
+      }
+    }
+  }
+
+  def postMessage(message: RequestMessage, p: Promise[Any]): Unit = {
+    val receiver = nameToEndpoint.get(message.receiver.name)
+    if (receiver != null) {
+      val inbox = endpointToInbox.get(receiver.endpoint)
+      if (inbox != null) {
+        val rpcCallContext =
+          new LocalNettyRpcCallContext(
+            inbox.endpointRef, message.senderAddress, message.needReply, p)
+        postMessageToInbox(inbox,
+          ContentMessage(message.senderAddress, message.content, message.needReply, rpcCallContext))
+      }
+    }
+  }
+
+  private def postMessageToInbox(inbox: Inbox, message: InboxMessage): Unit = {
+    inbox.post(message)
+    afterUpdateInbox(inbox)
+  }
+
+  private def afterUpdateInbox(inbox: Inbox): Unit = {
+    // Do some work to trigger processing messages in the inbox
+    val endpoint = inbox.endpoint
+    // Replacing unsuccessfully means someone is processing it
+    idleInboxes.replace(endpoint, inbox, inbox)
+    receivers.put(endpoint)
+  }
+
+  private[netty] class MessageLoop extends Runnable {
+    override def run(): Unit = {
+      try {
+        while (!stopped) {
+          try {
+            val endpoint = receivers.take()
+            val inbox =
+              if (endpoint.isInstanceOf[ThreadSafeRpcEndpoint]) {
```

Comment:
I wonder if, instead of this complicated handling of `idleInboxes`, if it wouldn't be simpler to have the inbox keep track of whether it's busy?

e.g. `ConcurrentInbox` would never be busy, but `ThreadSafeInbox` would become busy as soon as `process` is called, so other threads trying to call `process` would see that and could move on.